Create README.md (#68)
ossc-db/pg_hint_plan: README.md

| `application_name`  | The value of `application_name` of sessions to apply the hint. The hint in the example below applies to sessions connected from psql. An empty string means sessions of any `application_name`. |
| `hints`             | Hint phrase. This must be a series of hints excluding surrounding comment marks.                                                                                                            |

The following example shows how to operate with the hint table.

    postgres=# INSERT INTO hint_plan.hints(norm_query_string, application_name, hints)
    postgres-#     VALUES (
    postgres(#         'EXPLAIN (COSTS false) SELECT * FROM t1 WHERE t1.id = ?;',
    postgres(#         '',
    postgres(#         'SeqScan(t1)'
    postgres(#     );
    INSERT 0 1
    postgres=# UPDATE hint_plan.hints
    postgres-#    SET hints = 'IndexScan(t1)'
    postgres-#  WHERE id = 1;
    UPDATE 1
    postgres=# DELETE FROM hint_plan.hints
    postgres-#  WHERE id = 1;
    DELETE 1
    postgres=#

> The hint table is owned by the creator user and having the default previledges at the time of creation. during `CREATE EXTENSION`. Table hints are prioritized than comment hits.

Reviewer: ```suggestion
The hint table is owned by the creator user and having the default privileges at the time of creation. during `CREATE EXTENSION`. Table hints are prioritized over comment hits.
```

### The types of hints

> Hinting phrases are classified into six types based on what kind of object and how they can affect planning. Scaning methods, join methods, joining order, row number correction, parallel query and GUC setting. You will see the lists of hint phrases of each type in [Hint list](#hints-list).

Reviewer: ```suggestion
Hinting phrases are classified into six types based on what kind of object and how they can affect planning. Scanning methods, join methods, joining order, row number correction, parallel query, and GUC setting. You will see the lists of hint phrases of each type in [Hint list](#hints-list).
```

|:-------------------------------|:----------------------------------------------------------------------------------------------------------------------|:----------|
| `pg_hint_plan.enable_hint`       | True enbles `pg_hint_plan`.                                                                                         | `on`      |
| `pg_hint_plan.enable_hint_table` | True enbles hinting by table. `true` or `false`.                                                                    | `off`     |
| `pg_hint_plan.parse_messages`    | Specifies the log level of hint parse error. Valid values are `error`, `warning`, `notice`, `info`, `log`, `debug`. | `INFO`    |
| `pg_hint_plan.debug_print`       | Controls debug print and verbosity. Valid vaiues are `off`, `on`, `detailed` and `verbose`.                         | `off`     |
| `pg_hint_plan.message_level`     | Specifies message level of debug print. Valid values are `error`, `warning`, `notice`, `info`, `log`, `debug`.      | `INFO`    |

## Installation

This section describes the installation steps.

### building binary module

> Simply run `make` in the top of the source tree, then `make install` as appropriate user. The `PATH` environment variable should be set properly for the target PostgreSQL for this process.

Reviewer: ```suggestion
Simply run `make` at the top of the source tree, then `make install` as an appropriate user. The `PATH` environment variable should be set properly for the target PostgreSQL for this process.
```

    postgres=# CREATE VIEW v1 AS SELECT * FROM t2;
    postgres=# EXPLAIN /*+ HashJoin(t1 v1) */
              SELECT * FROM t1 JOIN v1 ON (c1.a = v1.a);
                                QUERY PLAN
    ------------------------------------------------------------------
     Hash Join  (cost=3.27..18181.67 rows=101 width=8)
       Hash Cond: (t1.a = t2.a)
       ->  Seq Scan on t1  (cost=0.00..14427.01 rows=1000101 width=4)
       ->  Hash  (cost=2.01..2.01 rows=101 width=4)
             ->  Seq Scan on t2  (cost=0.00..2.01 rows=101 width=4)

#### Inheritance tables

Hints can point only the parent of an inheritance tables and the hint affect all the inheritance. Hints simultaneously point directly to children are not in effect.

#### Hinting on multistatements

One multistatement can have exactly one hint comment and the hints affects all of the individual statement in the multistatement. Notice that the seemingly multistatement on the interactive interface of psql is internally a sequence of single statements so hints affects only on the statement just following.

#### VALUES expressions

> `VALUES` expressions in `FROM` clause are named as `*VALUES*` internally so it is hintable if it is the only `VALUES` in a query. Two or more `VALUES` expressions in a query seems distinguishable looking its explain result. But in reality it is mere a cosmetic and they are not distinguisable.

Reviewer: ```suggestion
`VALUES` expressions in `FROM` clause are named as `*VALUES*` internally so it is hintable if it is the only `VALUES` in a query. Two or more `VALUES` expressions in a query seem distinguishable looking at its explain result. But in reality, it is merely a cosmetic and they are not distinguishable.
```

#### Using `IndexOnlyScan` hint

Index scan may unexpectedly performed on another index when the index specifed in IndexOnlyScan hint cannot perform index only scan.

#### Behavior of `NoIndexScan`

`NoIndexScan` hint involes `NoIndexOnlyScan`.

#### Parallel hint and `UNION`

A `UNION` can run in parallel only when all underlying subqueries are parallel-safe. Conversely enforcing parallel on any of the subqueries let a parallel-executable `UNION` run in parallel. Meanwhile, a parallel hint with zero workers hinhibits a scan from executed in parallel.

#### Setting `pg_hint_plan` parameters by Set hints

`pg_hint_plan` parameters change the behavior of itself so some parameters doesn't work as expected.

-   Hints to change `enable_hint`, `enable_hint_tables` are ignored even though they are reported as "used hints" in debug logs.
-   Setting `debug_print` and `message_level` works from midst of the processing of the target query.

## Errors

`pg_hint_plan` stops parsing on any error and uses hints already parsed on the most cases. Following are the typical errors.

#### Syntax errors

> Any syntactical errors or wrong hint names are reported as an syntax error. These errors are reported in the server log with the message level which specified by pg\_hint\_plan.message\_level if pg\_hint\_plan.debug\_print is on and aboves.

Reviewer: ```suggestion
Any syntactical errors or wrong hint names are reported as a syntax error. These errors are reported in the server log with the message level specified by pg\_hint\_plan.message\_level if pg\_hint\_plan.debug\_print is on and above.
```

#### Object misspecifications

> Object misspecifications results silent ingorance of the hints. This kind of error is reported as "not used hints" in the server log by the same condtion to syntax errors.

Reviewer: ```suggestion
Object misspecifications result in silent ignorance of the hints. This kind of error is reported as "not used hints" in the server log by the same condition as syntax errors.
```

#### Nested comments

Hint comment cannot include another block comment within. If `pg_hint_plan` finds it, differently from other erros, it stops parsing and abandans all hints already parsed. This kind of error is reported in the same manner as other errors.

## Functional limitations

#### Influences of some of planner GUC parameters

The planner does not try to consider joining order for FROM clause entries more than `from_collapse_limit`. `pg_hint_plan` cannot affect joining order as expected for the case.

#### Hints trying to enforce unexecutable plans

Planner chooses any executable plans when the enforced plan cannot be executed.

-   `FULL OUTER JOIN` to use nested loop
-   To use indexes that does not have columns used in quals
-   To do TID scans for queries without ctid conditions

#### Queries in ECPG

ECPG removes comments in queries written as embedded SQLs so hints cannot be passed form those queries. The only exception is that `EXECUTE` command passes given string unmodifed. Please consider hint tables in the case.

#### Work with `pg_stat_statements`

> `pg_stat_statements` generates a query id ignoring comments. As the result the identical queires with different hints are summerized as the same query.

Reviewer: ```suggestion
`pg_stat_statements` generates a query id ignoring comments. As the result, the identical queries with different hints are summarized as the same query.
```
